File: backend/middleware/auth_middleware.py
"""JWT auth + RBAC dependencies for FastAPI routes."""
from __future__ import annotations
from fastapi import Depends, HTTPException, Request, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
import jwt

from utils.jwt_utils import decode_access_token
from services.user_service import get_user_by_id

bearer_scheme = HTTPBearer(auto_error=False)


async def require_auth(
    request: Request,
    creds: HTTPAuthorizationCredentials | None = Depends(bearer_scheme),
):
    if creds is None or creds.scheme.lower() != "bearer":
        raise HTTPException(status.HTTP_401_UNAUTHORIZED, "Missing bearer token")
    try:
        payload = decode_access_token(creds.credentials)
    except jwt.ExpiredSignatureError:
        raise HTTPException(status.HTTP_401_UNAUTHORIZED, "Access token expired")
    except jwt.InvalidTokenError:
        raise HTTPException(status.HTTP_401_UNAUTHORIZED, "Invalid access token")

    if payload.get("type") != "access":
        raise HTTPException(status.HTTP_401_UNAUTHORIZED, "Wrong token type")

    user = await get_user_by_id(payload["sub"])
    if not user or not user.is_active:
        raise HTTPException(status.HTTP_401_UNAUTHORIZED, "User not found or disabled")

    request.state.user = user
    return user


def require_roles(*roles: str):
    async def _dep(user=Depends(require_auth)):
        if not set(roles).intersection(user.roles):
            raise HTTPException(status.HTTP_403_FORBIDDEN, f"Requires role: {roles}")
        return user

    return _dep


async def _decode_optional(request: Request) -> str | None:
    """Extract user_id from a Bearer token if present. Returns None otherwise. Never raises."""
    auth = request.headers.get("authorization") or request.headers.get("Authorization")
    if not auth or not auth.lower().startswith("bearer "):
        return None
    token = auth.split(" ", 1)[1].strip()
    try:
        payload = decode_access_token(token)
        if payload.get("type") != "access":
            return None
        return payload.get("sub")
    except Exception:  # noqa: BLE001
        return None

File: backend/src/middleware/auth.middleware.js
const jwt = require('jsonwebtoken');
const ApiError = require('../utils/ApiError');
const { decodeAccessToken } = require('../utils/jwt.utils');
const config = require('../config');
const User = require('../models/User');

/**
 * Require authentication middleware.
 * Supports both JWT token payload formats (sub / userId, jwtSecret / jwt.accessSecret).
 * Extracts JWT from Authorization header or cookies, verifies it, and attaches user to req.
 */
const requireAuth = async (req, res, next) => {
  try {
    let token;
    const authHeader = req.headers.authorization;
    if (authHeader && authHeader.startsWith('Bearer ')) {
      token = authHeader.split(' ')[1];
    } else if (req.cookies?.accessToken) {
      token = req.cookies.accessToken;
    }

    if (!token) {
      throw ApiError.unauthorized('Missing bearer token');
    }

    let payload;
    try {
      payload = decodeAccessToken(token);
    } catch (err) {
      try {
        payload = jwt.verify(token, config.jwt.accessSecret);
      } catch (err2) {
        if (err.name === 'TokenExpiredError' || err2.name === 'TokenExpiredError') {
          throw ApiError.unauthorized('Access token expired');
        }
        throw ApiError.unauthorized('Invalid access token');
      }
    }

    const userId = payload.sub || payload.userId;
    if (!userId) {
      throw ApiError.unauthorized('Invalid token payload');
    }

    const user = await User.findById(userId);
    if (!user || user.is_active === false || user.isDeleted === true) {
      throw ApiError.unauthorized('User not found or disabled');
    }

    req.user = user;
    next();
  } catch (error) {
    next(error);
  }
};

/**
 * Optional auth — extracts user_id from token if present, does not throw.
 */
const optionalAuth = async (req, res, next) => {
  try {
    let token;
    const authHeader = req.headers.authorization;
    if (authHeader && authHeader.startsWith('Bearer ')) {
      token = authHeader.split(' ')[1];
    } else if (req.cookies?.accessToken) {
      token = req.cookies.accessToken;
    }
    if (!token) {
      req.userId = null;
      return next();
    }
    let payload;
    try {
      payload = decodeAccessToken(token);
    } catch {
      try {
        payload = jwt.verify(token, config.jwt.accessSecret);
      } catch {
        payload = null;
      }
    }
    req.userId = payload ? (payload.sub || payload.userId) : null;
  } catch {
    req.userId = null;
  }
  next();
};

module.exports = { requireAuth, optionalAuth };
